### ISSUE TITLE
New request/offer: splitting/fixing devdocs.io

### ISSUE THREAD

**Issue by josephlewis42** [openzim/zim-requests#1086]
- Website URL: https://devdocs.io
- License: Various open licenses
- Desired ZIM Title: Multiple?
- Language (ISO 639-3): eng
- Is this a MediaWiki?: no

Hi folks, I'm interested at getting some more programming content in Kiwix. [Devdocs has similar goals to this project](https://devdocs.io/) where they scrape something like 100 documentation sites, process the HTML, and bundle it in `localStorage` so it can be accessed offline.

I know there is a devdocs integration: https://github.com/openzim/zim-requests/issues/181, but it seems to be producing files that aren't passing validation. I think it would be worth splitting the site into multiple zims based on tool and/or tool and version as well e.g. `devdocs.io - Python 3.11`.

I'm happy to help do the work if you all think this is worthwhile and can point me in the right direction!

**Comment by benoit74** [comment]
Hi Joseph, thank you for the heads up!

I had a look at devdocs recipe and I see no reason it should not work ... but obviously there are plenty.

We've just released Zimit 2 which means software has changed quite a lot since the devdocs recipe was failing few month ago, so I've just restarted the recipe with a limit of first 20 pages just to see how the website works with Zimit 2. Should it be ok, I will request the full crawl.

I hope / think it could be fixable, and I unfortunately don't think that splitting the website would help to have the recipe succeed (aside if the fact they now have more and more languages tend to become an issue, but maintaining hundreds of recipes and ZIM files is going to be a hassle).

**Comment by benoit74** [comment]
Website is just not working inside the ZIM. Probably related to the fact they are using a service worker, which are not (yet) supported by zimit (I should have checked for this before, since you spoke about local storage ...), see https://github.com/openzim/warc2zim/issues/231. Investigation would probably help a bit, because it should have degraded to no service worker.

**Comment by josephlewis42** [comment]
Thanks for taking a look, @benoit74!

> I hope / think it could be fixable, and I unfortunately don't think that splitting the website would help to have the recipe succeed (aside if the fact they now have more and more languages tend to become an issue, but maintaining hundreds of recipes and ZIM files is going to be a hassle).

**Comment by benoit74** [comment]
> Are recipes a 1:1 in the farm? I'd be willing to build a new scraper if it made sense with the goals of the project to ETL this data into a bunch of independently usable Zims.

One single recipe can create multiple ZIM, this is typically the case with many mwoffliner recipes where we create 2 or 3 files per recipe (the different flavours), see e.g. https://farm.openzim.org/pipeline/34684b19-eb21-48e4-9975-61b034cbc4fd or we even have the gutenberg scraper which is creating 56 ZIMs at once.

I'm glad you are willing to help, and I think devdocs is indeed a good case for creating such a scraper. @rgaudin @kelson42 any PoV to share on this?

I think we should find the proper way to bootstrap you, we still miss much documentation on how to create a scraper from scratch, not speaking about the existence of a framework and/or tutorial.

Regarding details on how the scraper could work, I'm not yet sure what is the good strategy, but this can definitely be discussed later.

**Comment by rgaudin** [comment]
Indeed, there is no scraper howto. You should find [one of our scraper](https://github.com/openzim/overview/wiki/Scrapers-list) which looks like what you want to achieve and replicate.

**Comment by benoit74** [comment]
@josephlewis42 we've discussed briefly about your proposition yesterday with the team and we all agreed it makes sense. If it is ok for you, I'll setup a `devdocs` repo for you in openzim organization, and grant you maintainers right on this repo so that you're quite "free" on this repo.

My recommendation is to write the scraper in Python since it is now the main language used for scrapers, and the one where we have at the same time the most experience and most tooling, guidelines, ...

If you're ok with it, I would like that we use this opportunity to document how to create a scraper in our wiki. I've started to build a first version of this [howto](https://wiki.openzim.org/wiki/How-to_create_a_Python_scraper) in our wiki. Feel free to contribute to this, I've probably forgotten multiple things and it is still very minimalist.

**Comment by josephlewis42** [comment]
Hi @benoit74, that would be awesome!

I started a skeleton that allows me to pull down all the relevant docs but haven't committed anything yet. I'm very happy to contribute to the documentation as I go along. 😃 

Also, where is the best place to get in touch with the dev team, is it IRC? I've also seen that there's a mailing list and Mastodon links. I'm on the weest coast of the United States so I suspect there might be limited time zone overlap.

**Comment by kelson42** [comment]
@benoit74 Great, please join Slack #scraper channel! https://join.slack.com/t/kiwixoffline/shared_invite/zt-19s7tsi68-xlgHdmDr5c6MJ7uFmJuBkg

**Comment by Popolechien** [comment]
@kelson42 I think you meant to invite @josephlewis42 😁

**Comment by benoit74** [comment]
@Popolechien @kelson42 

We are now at the stage where a `0.1.0` of a custom `devdocs` scraper is going to be published in the coming hours/days.

This scraper will need one recipe per ZIM, each ZIM corresponding to one slug on devdocs.io

For instance we can create a ZIM for `dom` slug, corresponding to https://devdocs.io/dom/ (Web APIs).

Note that in many cases the slugs are versioned, e.g we have `angular` (most recent version) but also `angular~14`,  `angular~15`, ...

Which ZIM/slug do we want to create first? The idea would be to give it a try on a bunch of useful slugs, confirm quality is sufficient in dev.library.kiwix.org (or identify required issues before release to prod), and then decide on how to move forward (automatically create all recipes for all slugs, or I don't know what).

**Comment by Popolechien** [comment]
I don't know how much you need, but how about CSS and Python ? 

I see that CSS has a bunch of fun URLs like https://devdocs.io/css/::-webkit-scrollbar (with two colons), and python is overall usefull so we might as well have something popular that will draw in a few eyeballs.

**Comment by josephlewis42** [comment]
Hi @benoit74! I think it's ready for a v0.1.0 release.

I know there's a request for R (or at least CRAN) #1070, Vue.js #982, and C/C++ #953. The people who opened those might be willing to beta-test because they're already invested enough to open a bug.

**Comment by benoit74** [comment]
I've created and requested first recipes, let's see how it goes: https://farm.openzim.org/recipes?category=devdocs

**Comment by benoit74** [comment]
4 ZIMs are now ready at https://dev.library.kiwix.org/#lang=&tag=devdocs

Please review and do not hesitate to report issues on https://github.com/openzim/devdocs/ exceptionally (because the scraper is new so we might have many, we will triage the issues in this repo directly)

Note that lua is a single page (and hence has mostly no search result) because it is a single page in DevDocs.

**Comment by josephlewis42** [comment]
I pulled down C++, Lua, and Vue locally and reported a few issues with navigation and code syntax highlighting -- otherwise they seemed okay to me.

**Comment by josephlewis42** [comment]
@benoit74 with the v0.2.0 release, do you think that a subset of these are in a good enough spot to be published?

v0.3.0 is shaping up to be quite a bit of work so I'd like to make sure it's being put toward something that's useful for the community.

**Comment by benoit74** [comment]
@josephlewis42 thank you for confirming that 0.3.0 is quite a bit of work. We started to discuss what should be done in https://github.com/openzim/zim-requests/issues/1230. I will update it with a new proposition.

**Comment by benoit74** [comment]
With all files in prod, I think we can close this issue. Rest of the work is now normal gradual scraper enhancements, no need to track them here and in scraper repo.

**Comment by Popolechien** [comment]
I know we're moving to another tagging / categorization structure, but seeing as there is 200+ of them wouldn't it be worth assigning them to a category on the library?

**Comment by benoit74** [comment]
There is already a tag ... No strong PoV on categories since they are useless from my PoV currently. And just realize there is no ZIM category for now on these ZIMs.

**Comment by Popolechien** [comment]
As a side note should not the zim description be a bit more formal than `X docs by devdocs`? 

How about `X documentation, by DevDocs`

Again, seeing the number of recipes, only if this can be done by a quick script rather than manually.

**Comment by benoit74** [comment]
All this is automated, and quite very small changes, but there is nothing like less than 10' because we need to test scraper again, release scraper again, request recipes again, check ZIMs again. And this is exactly why we have a review process where we ask ourselves if the scraper / ZIM is ok before pushing to "the masses / prod", because late modifications cost a lot because you have to redo so many things. Unfortunately here we are too late.

So please open two issues so that we have a chance to include this in next iteration:
- one in devdocs repo to ask for a category
- one in zim-requests repo to ask for reconfiguration of the devdocs recipes description (and mention that we need to be sure it does not exceeds the 80 chars limit for any ZIM, I'm pretty sure it is OK but we are better safe than sorry)

**Comment by Popolechien** [comment]
Tags will do, they're only for comms purpose AFAIAC (as in, point at the tag in the announcement).

I'll open a request for the description but there's no rush and this can wait for the next automatic update in X months.

**Comment by benoit74** [comment]
OK, then I will update recipe configuration immediately so that it is not forgotten before next automatic update
